rfcs/0000-error-handling.md

```diff
@@ -0,0 +1,144 @@
+
+## Urgency
+
+Sooner is better than later.
+
+## Terminology
+
+No new terminology, but the semantics of the "Failed" status are altered.
+
+## Detailed Design
+
+### Error handling
+
+If a WASM handler terminates with a trap, that is an error. The node discards all entity operations and data source operations done by the handler and continue syncing.
+
+For subgraphs that are not yet synced, any handler error halts syncing, so errors in pending versions will not make their way to the current version. 
+
+When the node starts, we currently reset the failed flag on subgraphs. The flag is no longer be reset on synced subgraphs, since the subgraph will already be past the errors.
+
+Any graphql queries to a failed subgraph will include an error of the form:
+
+```
+{
+	message: "Subgraph failed"
+}
+```
+
+So that applications will always be aware that queries may have inconsistent data. Developers should refer to the indexing status API for details on the errors.
+
+### Indexing Status API
+
+The indexing status API will be revamped.
+
+The subgraph indexing status API currently has the top level query:
+
+    indexingStatusesForSubgraphName(subgraphName: String!): [SubgraphIndexingStatus!]!
+
+This is changed to:
+
+    indexingStatusesForSubgraphName(subgraphName: String!): SubgraphIndexingStatus
+
+Which returns null if the name is not found, and returns the status for the current version of the subgraph otherwise.
+
+A field `failedAfter`  is added to record the block previous to the first error.
+
+Errors are made more than just a string with the `Error` type.
+
+The complete indexing status API schema is:
+
+```graphql
+type Query {
+  indexingStatusesForSubgraphName(subgraphName: String!): SubgraphIndexingStatus
+  indexingStatuses(subgraphs: [String!]): [SubgraphIndexingStatus!]!
+}
+
+type SubgraphIndexingStatus {
+  subgraph: String!
+  synced: Boolean!
+  errors: [Error!]! # Sorted from first to last
+  chains: [ChainIndexingStatus!]!
+  node: String!
+}
+
+# Fields are moved from `EthereumIndexingStatus` to the interface.
+interface ChainIndexingStatus {
+  network: String!
+  chainHeadBlock: Block
+  earliestBlock: Block
+  latestBlock: Block
+  failedAfter: Block
+}
+
+type EthereumIndexingStatus implements ChainIndexingStatus { }
+
+# Renamed from `EthereumBlock` to `Block`
+type Block {
+  hash: Bytes!
+  number: BigInt!
+}
+
+type Error {
+	message: String!
+	
+	# Context for the error.
+	network: String!
+	block: Block!
+	transaction: Bytes
+	handler: String
+}
+```
+
+Since the indexing API is a view to the subgraph metadata, an `Error` type is also added to the metadata where subgraph errors are recorded.
+
+## Compatibility
+
+This does incompatible changes to the indexing status API, however that API was so far experimental and undocumented.
+
+The behavior of synced, failed subgraphs is changed to keep indexing. Our consensus is that the current behavior is useful to nobody and this can only be an improvement.
```

Comment:
I wonder if `failed` is still the right term given the new behavior. If subgraphs keep going, `failing` or `unhealthy` could be better, more clear terms. What do you think?

Reply:
I've done a major change to the RFC to explicitly use a third "unhealthy" state for this.